### PR TITLE
WIP: psf_ctf module revision

### PR DESCRIPTION
**Pull request by olafhauk**
The psf_ctf module has been substantially revised, especially with respect to the combination of sensor types via SVD. The new functions that perform SVDs for sub-leadfields and sub-inverse-matrices (within labels) separately for sensor types are _get_svd_comp() and _label_svd(). Feedback on all levels would be greatly appreciated.

### CODE REVIEW

File: mne/minimum_norm/psf_ctf.py [hunk 1]
```diff
@@ -8,7 +8,10 @@
 import numpy as np
 from scipy import linalg
 
-from ..io.pick import pick_channels
+from mne import pick_types, pick_info
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: relative import here, probably as `from mne.io.pick import ...`

File: mne/minimum_norm/psf_ctf.py [hunk 2]
```diff
@@ -8,7 +8,10 @@
...
+from .. import pick_channels_forward
+from ..io.pick import pick_channels, pick_types
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: actually these three lines can probably be combined

File: mne/minimum_norm/psf_ctf.py [hunk 3]
```diff
@@ -18,7 +21,7 @@
 
 
 def _prepare_info(inverse_operator):
-    """Get a usable dict."""
+    """Helper to get a usable dict"""
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: Undo this change. Docstrings should be in imperative voice

File: mne/minimum_norm/psf_ctf.py [hunk 4]
```diff
@@ -29,18 +32,16 @@ def _prepare_info(inverse_operator):
 
 
 def _pick_leadfield(leadfield, forward, ch_names):
-    """Pick out correct lead field components."""
-    # NB must pick from fwd['sol']['row_names'], not ['info']['ch_names'],
-    # because ['sol']['data'] may be ordered differently from functional data
-    picks_fwd = pick_channels(forward['sol']['row_names'], ch_names)
+    """Helper to pick out correct lead field components"""
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: Same

File: mne/minimum_norm/psf_ctf.py [hunk 5]
```diff
@@ -29,18 +32,16 @@ def _prepare_info(inverse_operator):
...
+    picks_fwd = pick_channels(forward['info']['ch_names'], ch_names)
     return leadfield[picks_fwd]
 
 
 @verbose
 def point_spread_function(inverse_operator, forward, labels, method='dSPM',
                           lambda2=1 / 9., pick_ori=None, mode='mean',
                           n_svd_comp=1, verbose=None):
-    """Compute point-spread functions (PSFs) for linear estimators.
+    """Compute point-spread functions (PSFs) for linear estimators
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: Revert, docstrings should end with a period

File: mne/minimum_norm/psf_ctf.py [hunk 6]
```diff
@@ -78,8 +79,7 @@ def point_spread_function(inverse_operator, forward, labels, method='dSPM',
         (irrelevant for 'sum' and 'mean'). Explained variances within
         sub-leadfields are shown in screen output.
     verbose : bool, str, int, or None
-        If not None, override default verbose level (see :func:`mne.verbose`
-        and :ref:`Logging documentation <tut_logging>` for more).
+        If not None, override default verbose level (see mne.verbose).
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: revert

File: mne/minimum_norm/psf_ctf.py [hunk 7]
```diff
@@ -90,97 +90,74 @@ def point_spread_function(inverse_operator, forward, labels, method='dSPM',
...
+        channel types, dictionaries may contain keys 'mag', 'grad' or 'eeg'.
     evoked_fwd : Evoked
         Forward solutions corresponding to PSFs in stc_psf
         If mode='svd': n_svd_comp components per label are created
         (i.e. n_svd_comp successive time points in mne_analyze)
         The last sample is the summed forward solution across all labels
         (sum is taken across summary measures).
     """
+
+    print "Point Spread Function!"
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: revert

File: mne/minimum_norm/psf_ctf.py [hunk 8]
```diff
@@ -90,97 +90,74 @@ def point_spread_function(inverse_operator, forward, labels, method='dSPM',
...
     if mode not in ['mean', 'sum', 'svd']:
         raise ValueError("mode must be 'svd', 'mean' or 'sum'. Got %s."
                          % mode)
 
     logger.info("About to process %d labels" % len(labels))
 
-    forward = convert_forward_solution(forward, force_fixed=False,
+    # CHANGED (PSF and CTF differ with respect to source orientations?)
+    # for forward computations, orientations will be fixed
+    forward = convert_forward_solution(forward, force_fixed=True,
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: Do we need to do it this way? This will not work for volume source spaces.

File: mne/minimum_norm/psf_ctf.py [hunk 9]
```diff
@@ -90,97 +90,74 @@ def point_spread_function(inverse_operator, forward, labels, method='dSPM',
...
+    ch_bads = [c for c in info_inv['bads'] if (c in inv_ch_names)]
+    forward['info']['bads'] = ch_bads
+    inverse_operator['info']['bads'] = ch_bads
 
-        # get vertices in source space inside label
-        fwd_idx = np.searchsorted(forward['src'][this_hemi]['vertno'], idx)
+    # CHANGED
+    source_ori = {FIFF.FIFFV_MNE_UNKNOWN_ORI: 'Unknown',
+                  FIFF.FIFFV_MNE_FIXED_ORI: 'Fixed',
+                  FIFF.FIFFV_MNE_FREE_ORI: 'Free'}
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: unused?

File: mne/minimum_norm/psf_ctf.py [hunk 10]
```diff
@@ -241,14 +219,16 @@ def _get_matrix_from_inverse_operator(inverse_operator, forward, labels=None,
...
+    label_singvals : list of dict
+        Singular values of SVD for sub-matrices of inverse operator
+        (only if mode='svd').
+        Diffent list entries per label. Since SVD is applied separately for
+        channel types, dictionaries may contain keys 'mag', 'grad' or 'eeg'.
     """
     mode = mode.lower()
 
+    # apply_inverse cannot produce 3 separate orientations
+    # therefore 'force_fixed=True' is required
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: This is true (for surface source spaces anyway) only until #3842 lands. Maybe it's worth waiting for that?

File: mne/minimum_norm/psf_ctf.py [hunk 11]
```diff
@@ -261,33 +241,60 @@ def _get_matrix_from_inverse_operator(inverse_operator, forward, labels=None,
...
     # free orientation constraint not possible because apply_inverse would
-    # combined components
-    invmat_mat_op = apply_inverse(ev_id, inverse_operator, lambda2=lambda2,
-                                  method=method, pick_ori='normal')
+    # combine components
 
-    logger.info("Dimension of inverse matrix: %s" % str(invmat_mat_op.shape))
+    # pick_ori='normal' required because apply_inverse won't give separate
+    # orientations
+    if ~inverse_operator['source_ori'] == FIFF.FIFFV_MNE_FIXED_ORI:
```
Comment (Eric89GXL) [inline mne/minimum_norm/psf_ctf.py]: if a != b

**Comment by olafhauk** [comment]
Thanks a lot for your comments, I'll work on them soon.
I'm happy to include vector source estimates when they are ready. This new version is already an important update of the old one, which computed SVD across all sensors, not per sensor type. But if the vector source estimates are around the corner, it might be more efficient to wait.

**Comment by Eric89GXL** [comment]
@olafhauk is this a precursor to implementing RRMat / adaptive parcellations ... ? :)

**Comment by olafhauk** [comment]
Rezvan is using these functions for her parcellations. I don't know when she'll have the time to contribute her code to mne-python, it may appear as a separate package first. It would be great to have this functionality available.

**Comment by larsoner** [comment]
@olafhauk vector source estimates are now available if you have time to come back to this.

**Comment by larsoner** [comment]
I'll close this in favor of  #6479 / moving PSF/CTF stuff to the new module
